ios/browser/api/history/brave_history_api.mm

```diff
@@ -0,0 +1,229 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/ios/browser/api/history/brave_history_api.h"
+
+#include "base/bind.h"
+#include "base/strings/sys_string_conversions.h"
+#include "components/history/core/browser/history_service.h"
+#include "components/history/core/browser/history_types.h"
+#include "ios/chrome/browser/history/history_service_factory.h"
+#include "ios/chrome/browser/history/web_history_service_factory.h"
+#include "ios/web/public/thread/web_thread.h"
+#include "net/base/mac/url_conversions.h"
+#include "ui/base/page_transition_types.h"
+#include "url/gurl.h"
+
+#include "brave/ios/browser/api/history/brave_history_observer.h"
+#include "brave/ios/browser/api/history/history_service_listener_ios.h"
+
+#if !defined(__has_feature) || !__has_feature(objc_arc)
+#error "This file requires ARC support."
+#endif
+
+#pragma mark - IOSHistoryNode
+
+@interface IOSHistoryNode () {
+  std::u16string title_;
+  GURL gurl_;
+  base::Time date_added_;
+}
+@end
+
+@implementation IOSHistoryNode
+
+- (instancetype)initWithURL:(NSURL*)url
+                      title:(NSString* _Nullable)title
+                  dateAdded:(NSDate* _Nullable)dateAdded {
+  if ((self = [super init])) {
+    [self setUrl:url];
+
+    if (title) {
+      [self setTitle:title];
+    }
+
+    if (dateAdded) {
+      [self setDateAdded:dateAdded];
+    }
+  }
+
+  return self;
+}
+
+- (void)setUrl:(NSURL*)url {
+  gurl_ = net::GURLWithNSURL(url);
+}
+
+- (NSURL*)url {
+  return net::NSURLWithGURL(gurl_);
+}
+
+- (void)setTitle:(NSString*)title {
+  title_ = base::SysNSStringToUTF16(title);
+}
+
+- (NSString*)title {
+  return base::SysUTF16ToNSString(title_);
+}
+
+- (void)setDateAdded:(NSDate*)dateAdded {
+  date_added_ = base::Time::FromNSDate(dateAdded);
+}
+
+- (NSDate*)dateAdded {
+  return date_added_.ToNSDate();
+}
+@end
+
+#pragma mark - BraveHistoryAPI
+
+@interface BraveHistoryAPI () {
+  // History Service for adding and querying
+  history::HistoryService* history_service_;
+  // WebhistoryService for delete operations
+  history::WebHistoryService* web_history_service_;
+  // Tracker for history requests.
+  base::CancelableTaskTracker tracker_;
+}
+@property(nonatomic, strong) void (^query_completion)(NSArray<IOSHistoryNode*>*)
+    ;
+@end
+
+@implementation BraveHistoryAPI
+
+- (instancetype)initWithHistoryService:(history::HistoryService*)historyService
+                     webHistoryService:
+                         (history::WebHistoryService*)webHistoryService {
+  if ((self = [super init])) {
+    DCHECK_CURRENTLY_ON(web::WebThread::UI);
```

Comment:
nit: Shouldn't the DCHECK be outside the if block?

Reply:
The reason is `self = [super init]` is returning its own receiver (the self pointer doesn't change) with inherited instance values initialized.

In this case it is better to check which thread we are on while setting the services.